File: day4/go/main.go
package main

type MyCircularQueue struct {
	data     []int
	front    int
	rear     int
	size     int
	capacity int
}

func Constructor(k int) MyCircularQueue {
	return MyCircularQueue{
		data:     make([]int, k),
		capacity: k,
	}
}

func (q *MyCircularQueue) EnQueue(value int) bool {
	if q.IsFull() {
		return false
	}
	q.data[q.rear] = value
	q.rear = (q.rear + 1) % q.capacity
	q.size++
	return true
}

func (q *MyCircularQueue) DeQueue() bool {
	if q.IsEmpty() {
		return false
	}
	q.front = (q.front + 1) % q.capacity
	q.size--
	return true
}

func (q *MyCircularQueue) Front() int {
	if q.IsEmpty() {
		return -1
	}
	return q.data[q.front]
}

func (q *MyCircularQueue) Rear() int {
	if q.IsEmpty() {
		return -1
	}
	return q.data[(q.rear+q.capacity-1)%q.capacity]
}

func (q *MyCircularQueue) IsEmpty() bool {
	return q.size == 0
}

func (q *MyCircularQueue) IsFull() bool {
	return q.size == q.capacity
}

File: day4/typescript/main.ts
export class MyCircularQueue {
    private data: number[];
    private front: number;
    private rear: number;
    private size: number;
    private capacity: number;

    constructor(k: number) {
        this.data = new Array(k);
        this.front = 0;
        this.rear = 0;
        this.size = 0;
        this.capacity = k;
    }

    enQueue(value: number): boolean {
        if (this.isFull()) return false;
        this.data[this.rear] = value;
        this.rear = (this.rear + 1) % this.capacity;
        this.size++;
        return true;
    }

    deQueue(): boolean {
        if (this.isEmpty()) return false;
        this.front = (this.front + 1) % this.capacity;
        this.size--;
        return true;
    }

    Front(): number {
        return this.isEmpty() ? -1 : this.data[this.front];
    }

    Rear(): number {
        return this.isEmpty() ? -1 : this.data[(this.rear + this.capacity - 1) % this.capacity];
    }

    isEmpty(): boolean {
        return this.size === 0;
    }

    isFull(): boolean {
        return this.size === this.capacity;
    }
}
